src/next-app/pages/create-account.tsx

```diff
@@ -1,84 +1,132 @@
+import { registrationFn } from '../utils/api';
+import { useRouter } from 'next/router';
+
+type Field = {
+  value: string;
+  errors: string[];
+};
+
+function fieldIsValid(field: Field) {
+  return field.value.length > 0 && field.errors.length === 0;
+}
 
 const CreateAccountPage: NextPage = () => {
-  const [firstName, setFirstName] = useState('');
-  const [lastName, setLastName] = useState('');
-  const [phoneNumber, setPhoneNumber] = useState('');
-  const [email, setEmail] = useState('');
-  const [emailErrors, setEmailErrors] = useState<Set<string> | undefined>(undefined);
-  const [username, setUsername] = useState('');
-  const [password, setPassword] = useState('');
-  const valid =
-    firstName.length > 0 &&
-    lastName.length > 0 &&
-    phoneNumber.length > 0 &&
-    email.length > 0 &&
-    username.length > 0 &&
-    password.length > 0 &&
-    emailErrors === undefined;
+  const router = useRouter();
+  const [firstName, setFirstName] = useState<Field>({ value: '', errors: [] });
+  const [lastName, setLastName] = useState<Field>({ value: '', errors: [] });
+  const [phoneNumber, setPhoneNumber] = useState<Field>({ value: '', errors: [] });
+  const [email, setEmail] = useState<Field>({ value: '', errors: [] });
+  const [username, setUsername] = useState<Field>({ value: '', errors: [] });
+  const [password, setPassword] = useState<Field>({ value: '', errors: [] });
+  const fields = [firstName, lastName, phoneNumber, email, username, password];
+  const valid = fields.every(fieldIsValid);
 
-  const handleEmailBlur = () => {
-    if (email.length > 0 && !validateEmail(email)) {
-      setEmailErrors(new Set(['Invalid email format']));
-    } else {
-      setEmailErrors(undefined);
-    }
-  };
+  const mutation = useMutation({
+    mutationFn: registrationFn,
+  });
 
   const handleSubmit = () => {
-    window.location.href = '/invite';
+    mutation.mutate(
+      {
+        email: email.value,
+        username: username.value,
+        password: password.value,
+        re_password: password.value,
+        type: 'shop_owner',
+        first_name: firstName.value || undefined,
+        last_name: lastName.value || undefined,
+      },
+      {
+        onSuccess(data, variables, context) {
+          if (data.ok) {
+            router.push('/create-shop');
+          } else {
+            setEmail((prev) => {
+              return { ...prev, errors: [] };
+            });
+            data.json().then((response) => {
+              const setterMap = new Map<string, Dispatch<SetStateAction<Field>>>();
+              setterMap.set('email', setEmail);
+              setterMap.set('username', setUsername);
+              setterMap.set('password', setPassword);
+              for (let key of Object.keys(response)) {
+                const setter = setterMap.get(key);
+                setter &&
+                  setter((prev) => {
+                    return { ...prev, errors: response[key] };
+                  });
```

Comment:
Errors caught here are not being displayed to the user. For example, a duplicate username results in an error 400 but it's not being shown anywhere in the UI.

Reply:
Resolved. Updated logic for new api format.